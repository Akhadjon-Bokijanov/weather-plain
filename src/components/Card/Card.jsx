import React from 'react'

const Card = ({ img, min, max, temp, day, dt }) => {
    return (
            <div 
            data-testid="card-test-id"
            style={{ border: '1px #999 solid', borderRadius: 11, cursor: 'pointer', minHeight: 106, minWidth: 106, margin: 3}}>
            <div style={{padding: 8, display: 'flex', flexDirection: 'column', alignItems: 'center'}}>
                <div>
                    <strong>{day}</strong>
                </div>
                <div>
                    <img src={img} alt="weather img" width={60} height={60} />
                </div>
                {
                    temp ? <div style={{textAlign: 'center'}}>{temp}{'\u00b0'}</div>
                    : <div style={{ display: 'flex', justifyContent: 'space-between', }}>
                        <div>{min}{'\u00b0'}</div>
                        <div style={{ marginLeft: 10 }}>{max}{'\u00b0'}</div>
                    </div>
                }
                
            </div>
        </div>
    )
}

export default Card
