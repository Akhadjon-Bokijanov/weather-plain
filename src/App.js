import React, { useState, useEffect } from 'react';
import { Switch, Route, Link, withRouter } from 'react-router-dom';
import Card from './components/Card/Card';
import { API_KEY, LAT, LON } from './env';
import Hourly from './components/Hourly/Hourly';




function App() {

    const [zone, setZone] = useState('');
    const [data, setData] = useState();
    const [hData, setHData] = useState();
    const [hShown, setHShown] = useState();
    const [day, setDay] = useState();
    
    useEffect(() => {
        let p = new Promise((resolve, reject) => {
            resolve(
                fetch(`https://api.openweathermap.org/data/2.5/onecall?lat=${LAT}&lon=${LON}&exclude=minutely,current,alerts&appid=${API_KEY}`)
                    .then(res => res.json())
            );
        });
        p.then((data) => {
            const { daily, timezone, hourly } = data;
            setHData(hourly);
            setZone(timezone);
            console.log(timezone)

            //as just 5 days required
            setData(daily.filter((f, idx) => idx < 5).map(d => (
                {
                    dt: d.dt,
                    day: DayIndex[new Date(d.dt * 1000).getDay()].short,
                    min: d.temp.min,
                    max: d.temp.max,
                    //app should not crash when there is no weather back from response, and show clear day
                    img: `/assets/${d.weather[0]?.icon ?? '01d'}.png`
                }
            )));

        })
            .catch(e => {
                alert("Data is not fetched!");
                console.log(e)
            })
    }, []);


    //if day or data changes, we should change hourly weather data
    useEffect(() => {
        let dayIndex = DayIndex.findIndex(d => d.short === day);
        
        //find the dt of the day
        let id = data?.find(d => new Date(d.dt * 1000).getDay() === dayIndex)?.dt;

        if (id) {
            //set hourly data to show
            setHShown(hData.filter(item => parseInt(item.dt / (3600 * 24)) === parseInt(id / (3600 * 24))).map(f => (
                {
                    day: new Date(f.dt * 1000).getHours() + ":00",
                    dt: f.dt,
                    temp: f.temp,
                    img: `/assets/${f.weather[0]?.icon}.png`
                }
            )))
        }
    
    
    }, [day, data])

    return (
        <div className="App">
            <h3 style={{ textAlign: 'center' }}>Weather App: {zone}</h3>
            <Switch>
                <Route
                    key="daily"
                    path="/"
                    exact
                    render={() => (
                        <div
                            style={{ display: 'flex', justifyContent: 'space-around', width: '60%', margin: '30px auto' }}>
                            {
                                //when data is not fetched or still fetching, it must not crash
                                data?.map((item, idx) => <Link
                                    key={`link-${idx}`}
                                    to={`/${item.day}`}
                                    onClick={()=>setDay(item.day)}
                                    style={{ color: '#000', textDecoration: 'none' }}>
                                    {
                                        <Card {...item} key={idx} />
                                    }

                                </Link>) ?? 'Loading...'
                            }
                        </div>)} />
                <Route
                    key="hourly"
                    path="/:day"
                    exact
                    render={({ match: { params } }) => <Hourly
                        params={params}
                        hShown={hShown} />
                    } />


            </Switch>

        </div>
    );
}

export default App;

export const DayIndex = [
    { short: "Sun", long: "Sunday" },
    { short: "Mon", long: "Monday" },
    { short: "Thue", long: "Thuesday" },
    { short: "Wed", long: "Wednesday" },
    { short: "Thur", long: "Thursday" },
    { short: "Fri", long: "Friday" },
    { short: "Sat", long: "Saturday" },
]


export const sample = [
    {
        day: 'Wed',
        min: 78,
        max: 67,
        img: './cloud.png'
    },
    {
        day: 'Thu',
        min: 78,
        max: 67,
        img: './sunny.jpg'
    },
    {
        day: 'Fri',
        min: 78,
        max: 67,
        img: './cloud.png'
    },
    {
        day: 'Sat',
        min: 78,
        max: 67,
        img: './rainy.png'
    },
    {
        day: 'San',
        min: 78,
        max: 67,
        img: './snow.png'
    }
]

