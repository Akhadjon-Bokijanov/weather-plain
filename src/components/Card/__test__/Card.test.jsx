import { render, cleanup } from '@testing-library/react';
import React from 'react';
import ReactDOM from 'react-dom';
import Card from "../Card";
import { sample } from '../../../App';
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

afterEach(cleanup);

it("renders without issues", ()=>{
    const div = document.createElement('div');
    ReactDOM.render(<Card />, div);
});

it("render with styles", ()=>{
    let data = sample[0];
    const { getByTestId } = render(<Card { ...data } />)

    expect(getByTestId("card-test-id")).toHaveStyle("cursor: pointer");
})

it("snapshot test", ()=>{
    const s = renderer.create(<Card { ...sample[1] } />).toJSON();

    expect(s).toMatchSnapshot();
});