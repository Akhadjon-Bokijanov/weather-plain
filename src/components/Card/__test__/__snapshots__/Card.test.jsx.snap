// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  data-testid="card-test-id"
  style={
    Object {
      "border": "1px #999 solid",
      "borderRadius": 11,
      "cursor": "pointer",
      "margin": 3,
      "minHeight": 106,
      "minWidth": 106,
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "padding": 8,
      }
    }
  >
    <div>
      <strong>
        Thu
      </strong>
    </div>
    <div>
      <img
        alt="weather img"
        height={60}
        src="./sunny.jpg"
        width={60}
      />
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "justifyContent": "space-between",
        }
      }
    >
      <div>
        78
        °
      </div>
      <div
        style={
          Object {
            "marginLeft": 10,
          }
        }
      >
        67
        °
      </div>
    </div>
  </div>
</div>
`;
